runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -298,15 +298,25 @@ function M.on_diagnostic(_, result, ctx, config)
   local uri = ctx.params.textDocument.uri
   local fname = vim.uri_to_fname(uri)
 
-  if result == nil then
-    return
-  end
-
-  if result.kind == 'unchanged' then
-    return
+  local diagnostics = {}
+  if result ~= nil then
+    if result.kind == 'unchanged' then
+      return
+    end
+    diagnostics = result.items
+  else
+    local client_name = vim.lsp.get_client_by_id(client_id).name
+    vim.notify_once(
+      string.format(
+        "LSP: client '%s' (id %s) returned nil response to textDocument/diagnostic.\nThis is a bug in the LSP backend, and should be fixed within '%s'.",
```

Comment:
Nit but I find the `response` redundant.
```suggestion
        "LSP: client '%s' (id %s) returned nil to textDocument/diagnostic.\nThis is a bug in the LSP backend, and should be fixed within '%s'.",
```